### DISCUSSION TITLE
How is Zustand able to isolate re-renders of components to only changes of selected properties by said component

### DISCUSSION THREAD

**Discussion by Fawwaz-2009** [pmndrs/zustand#1732] [Q&A]
Hello all 
https://codesandbox.io/s/zustand-react-demo-forked-up3qvi?file=/src/components/ToggleTheme.js

I'm trying to understand how zustand is able to isolate re-renders in the way it does (see example above). I'm going off the custom hook mental model and following that if the hook re-render the component using the hook will re-render. That doesn't appear to be the case in Zustand. **To illustrate the issue I wrapped the Zustand store state in a custom hook and keeping pretty much the same selector api, and now the re-renders are not isolated to selected properties!**.

![image](https://user-images.githubusercontent.com/32019791/229671411-640dc99e-6b3a-44e7-9a6f-6e5f7c0ee8e7.png)

I'm aware that Zustand uses useSyncExternalStore, but from the react docs it doesn't appear that useSyncExternalStore should have any more features above what included in useState or useReducer  [Note Section from the docs](https://react.dev/reference/react/useSyncExternalStore#subscribing-to-an-external-store)

Here is the link for the new repo: https://codesandbox.io/s/zustand-react-demo-forked-rb0ogb?file=/src/stores/configStore.js

**Comment by JacobWeisenburger** [comment]
I think this is what you are looking for, but I'm not really sure.
https://github.com/pmndrs/zustand/blob/142b60025b41ed3407c2abb08afe3f048ea6258b/src/react.ts#L35-L49

https://codesandbox.io/s/zustand-react-demo-forked-ygt6ii?file=/src/stores/configStore.js

Let me know if you have any questions

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by Fawwaz-2009** [reply]
of course this works as you are just passing the selector to `configureStore`, but the questions is why it works 😄? . How can one create a similar functionality using `useState`?

**Reply by JacobWeisenburger** [reply]
I don't think you can do it with `useState`. I'm not sure how the internals of `useSyncExternalStore` work, but I'm pretty sure you have to use that.

perhaps something like this:
```ts
useSyncExternalStore(
    store.subscribe,
    () => selector ? selector( store.getState() ) : store.getState()
)
```

**Reply by Fawwaz-2009** [reply]
what is weird if you look at the documentation for [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore) it says that this only meant to make it easy to work with non react code and in most cases you should use `useState` or `useReducer` 🤔 

> When possible, we recommend using built-in React state with [useState](https://react.dev/reference/react/useState) and [useReducer](https://react.dev/reference/react/useReducer) instead. The useSyncExternalStore API is mostly useful if you need to integrate with existing non-React code.
